src/ripple/conditions/condition.h

```diff
@@ -0,0 +1,159 @@
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CONDITIONS_CONDITION_H
+#define RIPPLE_CONDITIONS_CONDITION_H
+
+#include <ripple/conditions/impl/base64.h> // use Beast implementation
+#include <ripple/conditions/impl/utils.h>
+#include <boost/optional.hpp>
+#include <array>
+#include <cstdint>
+#include <string>
+
+namespace ripple {
+namespace cryptoconditions {
+
+// NIKB-TODO: These should move to a separate file:
+// NIKB-TODO: These should be `enum class : std::uint16_t`
+std::uint16_t constexpr condition_hashlock         = 0;
+std::uint16_t constexpr condition_prefix_sha256    = 1;
+std::uint16_t constexpr condition_threshold_sha256 = 2;
+std::uint16_t constexpr condition_rsa_sha256       = 3;
+std::uint16_t constexpr condition_ed25519          = 4;
+
+// NIKB-TODO: These should be `enum class : std::uint32_t`
+std::uint32_t constexpr feature_sha256             = 1;
+std::uint32_t constexpr feature_preimage           = 2;
+std::uint32_t constexpr feature_prefix             = 4;
+std::uint32_t constexpr feature_threshold          = 8;
+std::uint32_t constexpr feature_rsa_pss            = 16;
+std::uint32_t constexpr feature_ed25519            = 32;
+
+struct condition_t
+{
+    std::uint16_t type;
+    std::uint16_t fulfillment_length;
+    std::uint32_t requires;
+    std::array<std::uint8_t, 32> fingerprint;
+
+    // Can this be deleted?
+    condition_t () = default;
+
+    condition_t (condition_t const&) = default;
+    condition_t (condition_t&&) = default;
+};
+
+inline
+bool
+operator== (condition_t const& lhs, condition_t const& rhs)
+{
+    return
+        lhs.type == rhs.type &&
+        lhs.requires == rhs.requires &&
+        lhs.fulfillment_length == rhs.fulfillment_length &&
+        lhs.fingerprint == rhs.fingerprint;
+}
+
+inline
+bool
+operator!= (condition_t const& lhs, condition_t const& rhs)
+{
+    return !(lhs == rhs);
+}
+
+/** Determine if a given condition is valid.
+
+    @note this function checks whether it understands the
+          type of the condition, and if so, whether it meets
+          the requirements mandated by the RFC.
+*/
+inline
+bool
+validate (condition_t const& c)
+{
+    // This check can never trigger because of the range of
+    // the fulfillment_length type. It's here in case the
+    // type is changed in the future.
+
+    // FIXME: maxFulfillmentLength
+    if (c.fulfillment_length > 65535)
+        return false;
+
+    if (c.type == condition_hashlock)
+        return (c.requires & (feature_sha256 | feature_preimage));
```

Comment:
I'm a little confused. So `condition_hashlock` needs to set these two features, but can also set, say `feature_ed25519` and it's still valid?